packages/fork-choice/src/protoArray/protoArray.ts

```diff
@@ -214,10 +237,245 @@ export class ProtoArray {
+          break;
+
+        default:
+      }
+
+      node = node.parent !== undefined ? this.getNodeByIndex(node.parent) : undefined;
+    }
+  }
+
+  /**
+   * Do a two pass invalidation:
+   *  1. we go up and mark all nodes invalid and then
+   *  2. we need do iterate down and mark all children of invalid nodes invalid
+   *
+   * latestValidHashIndex === undefined implies invalidate only invalidateTillIndex
+   * latestValidHashIndex === -1 implies invalidate all post merge blocks
+   * latestValidHashIndex >=0 implies invalidate the chain upwards from invalidateTillIndex
+   */
+
+  propagateInValidExecutionStatusByIndex(invalidateTillIndex: number, latestValidHashIndex: number | undefined): void {
+    let invalidateIndex: number = invalidateTillIndex;
+    // If latestValidHashIndex is undefined, i.e. we didn't find the LVH, only
+    // invalidate invalidateTillIndex
+    const invalidateUntilIndex = latestValidHashIndex ?? invalidateTillIndex - 1;
+
+    while (invalidateIndex > invalidateUntilIndex) {
+      const invalidNode = this.getNodeFromIndex(invalidateIndex);
+
+      if (invalidNode.executionStatus === ExecutionStatus.PreMerge) {
+        // This is also problematic case since lvh should have come as `0x00..00` with forkchoice
+        // passing us correct latestValidHashIndex. But we can be forgiving here and breakout
+        // if we reach pre-merge
+        break;
+      } else if (invalidNode.executionStatus === ExecutionStatus.Valid) {
+        // This is a catastrophe, and indicates consensus failure and a non
+        // recoverable damage. There is no further processing that can be done.
+        // Just assign error for marking proto-array perma damaged and throw!
+        this.lvhError = {
+          lvhCode: LVHExecErrorCode.ValidToInvalid,
+          blockRoot: invalidNode.blockRoot,
+          execHash: invalidNode.executionPayloadBlockHash,
+        };
+        throw new ProtoArrayError({
+          code: ProtoArrayErrorCode.INVALID_LVH_EXECUTION_RESPONSE,
+          ...this.lvhError,
+        });
+      }
+
+      invalidNode.executionStatus = ExecutionStatus.Invalid;
+      if (invalidNode.parent === undefined) {
+        // We have invalidated the entire chain in forkchoice, so time to exit and
+        // move to phase 2
+        break;
+      }
+      invalidateIndex = invalidNode.parent;
+    }
+
+    // Pass 2: mark all children of invalid nodes as invalid
+    for (let nodeIndex = 0; nodeIndex < this.nodes.length; nodeIndex++) {
+      const node = this.getNodeFromIndex(nodeIndex);
+      const parent = node.parent !== undefined ? this.getNodeByIndex(node.parent) : undefined;
+      // Only invalidate if this is post merge, and either parent is invalid or the
+      // concensus has failed
+      if (node.executionStatus !== ExecutionStatus.PreMerge && parent?.executionStatus === ExecutionStatus.Invalid) {
+        if (node.executionStatus === ExecutionStatus.Valid) {
```

Comment:
Should this two if statements be merged?

Reply:
simplified the conditions, hopefully its more clear now